### PR TITLE
doc, test: improve i2p/tor docs and i2p reachable unit tests

### PR DESCRIPTION
**Pull request by jonatack**
This pull addresses https://github.com/bitcoin/bitcoin/issues/22634#issuecomment-894104681 and various user feedback/questions, updates the -onlynet documentation in doc/i2p.md and doc/tor.md per #22651 (src/init.cpp is already fine) and fills in some missing I2P unit test coverage.

Note: this PR depends in part on whether #22651 is merged in order to propose the correct -onlynet documentation (it is currently aligned with the change in #22651), so that PR should be decided or merged first.

### CODE REVIEW

File: doc/i2p.md [hunk 1]
```diff
@@ -10,11 +10,20 @@ started with I2P terminology.
...
+  alternative in C++ (version 2.23 in Debian stable will not work; version 2.35
+  and up recommended)
+- [i2p-zero](https://github.com/i2p-zero/i2p-zero)
+- [other alternatives](https://en.wikipedia.org/wiki/I2P#Routers)
+
+Note the IP address and port the SAM proxy is listening to; usually, it is
+`127.0.0.1:7656`.
+
+Once it is up and running with SAM enabled, use the following Bitcoin Core
+configuration options:
```
Comment (jonatack) [inline doc/i2p.md:26]:
```diff
-Once it is up and running with SAM enabled, use the following Bitcoin Core
-configuration options:
+Once an I2P router with SAM enabled is up and running, use the following Bitcoin
+Core configuration options:
```

File: doc/i2p.md [hunk 2]
```diff
@@ -47,10 +56,13 @@ information in the debug log about your I2P configuration and connections. Run
 `bitcoin-cli help logging` for more information.
 
 It is possible to restrict outgoing connections in the usual way with
-`onlynet=i2p`. I2P support was added to Bitcoin Core in version 22.0 (mid-2021)
-and there may be fewer I2P peers than Tor or IP ones. Therefore, using
-`onlynet=i2p` alone (without other `onlynet=`) may make a node more susceptible
-to [Sybil attacks](https://en.bitcoin.it/wiki/Weaknesses#Sybil_attack). Use
+`onlynet=i2p`, combined with `-onion=0` if disabling outbound onion connections
+is desired, see [doc/tor.md](/tor.md).
```
Comment (jonatack) [inline doc/i2p.md:60]:
```suggestion
is desired, see [doc/tor.md](doc/tor.md).
```

File: doc/i2p.md [hunk 3]
```diff
@@ -10,11 +10,20 @@ started with I2P terminology.
...
-enabled is required (there is an [official one](https://geti2p.net) and
-[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP
-address and port the SAM proxy is listening to; usually, it is
-`127.0.0.1:7656`. Once it is up and running with SAM enabled, use the following
-Bitcoin Core options:
+enabled is required. Options include:
+
+- [i2p-router](https://geti2p.net), the official implementation in Java
+- [i2pd (I2P Daemon)](https://i2pd.readthedocs.io/en/latest), a lighter
+  alternative in C++ (version 2.23 in Debian stable will not work; version 2.35
```
Comment (jonatack) [inline doc/i2p.md:17]: I've tested 2.23 with both 22.0rc2 and master and it still manages to work, albeit with a few more warnings, some of which seemed to be false negatives before managing to connect, and the warnings when a peer isn't online are verbose and a bit less helpful:

```
2021-08-06T15:37:05Z I2P: Error connecting to zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p:0:
Unexpected reply to "NAMING LOOKUP NAME=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p":
"NAMING REPLY RESULT=INVALID_KEY NAME=zsxwyo6qcn3chqzwxnseusqgsnuw3maqnztkiypyfxtya4snkoka.b32.i2p"
```
Still testing but may just warn here instead.

File: doc/i2p.md [hunk 4]
```diff
@@ -47,10 +56,13 @@ information in the debug log about your I2P configuration and connections. Run
...
 
 It is possible to restrict outgoing connections in the usual way with
-`onlynet=i2p`. I2P support was added to Bitcoin Core in version 22.0 (mid-2021)
-and there may be fewer I2P peers than Tor or IP ones. Therefore, using
-`onlynet=i2p` alone (without other `onlynet=`) may make a node more susceptible
-to [Sybil attacks](https://en.bitcoin.it/wiki/Weaknesses#Sybil_attack). Use
+`onlynet=i2p` (combined with `-onion=0` if disabling outbound onion connections
+is desired; see [doc/tor.md](tor.md)).
+
+I2P support was added to Bitcoin Core in version 22.0 (mid-2021) and there may
```
Comment (kristapsk) [inline doc/i2p.md:62]: Mid-2021? 22.0 is not released yet.

Comment (jonatack) [reply doc/i2p.md:62]: 22.0 is the first release that `doc/i2p.md` will be in.

Comment (jonatack) [reply doc/i2p.md:62]: If the release takes place in August or early September, "mid-2021" seems ok to me but can change or remove if controverserial.  It wasn't added by this PR.

Comment (kristapsk) [reply doc/i2p.md:62]: Yes, I would not call September mid-2021, more like "autumn 2021", but OTOH I also not use where "mid-year" begins or ends (June or July would be definitely mid-2021 I think). Original schedule for 22.0 final release was August 1, but we are already past that.

File: doc/i2p.md (hunk 3 above)
Comment (jonatack) [reply doc/i2p.md:17]: done

File: doc/i2p.md (hunk 4 above)
Comment (jonatack) [reply doc/i2p.md:62]: Removed `(mid-2021)`

**Comment by prayank23** [comment]
Concept ACK on improving the doc.

File: doc/i2p.md [hunk 5]
```diff
@@ -42,15 +42,30 @@ named `i2p_private_key` in the Bitcoin Core data directory.
...
+
+```
+-onlynet=i2p
+```
+
+Make outgoing connections only to I2P addresses. Incoming connections are not
+affected by this option. It can be specified multiple times to allow multiple
+network types, e.g. onlynet=ipv4, onlynet=ipv6, onlynet=onion, onlynet=i2p.
+
+Warning: if you use -onlynet with values other than onion, and the -onion or
```
Comment (prayank23) [inline doc/i2p.md:72]: Warning looks okay or maybe we can wait for https://github.com/bitcoin/bitcoin/pull/22651 to get merged and update this accordingly

File: doc/i2p.md [hunk 6]
```diff
@@ -10,11 +10,22 @@ started with I2P terminology.
...
 A running I2P router (proxy) with [SAM](https://geti2p.net/en/docs/api/samv3)
-enabled is required (there is an [official one](https://geti2p.net) and
-[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP
-address and port the SAM proxy is listening to; usually, it is
-`127.0.0.1:7656`. Once it is up and running with SAM enabled, use the following
-Bitcoin Core options:
+
+enabled is required. Options include:
+
+- [i2p-router](https://geti2p.net), the official implementation in Java
```
Comment (prayank23) [inline doc/i2p.md:16]: Not sure if this implementation in Java should be mentioned because it doesn't help with Bitcoin Core setup

Comment (jonatack) [reply doc/i2p.md:16]: I'm not sure what you mean. This was already in the doc, and I've heard from people who are using it with Bitcoin Core.

File: doc/i2p.md (hunk 5 above)
Comment (jonatack) [reply doc/i2p.md:72]: My goal was to write this to be compatible with that PR.  Let me know if it isn't correct.

File: doc/i2p.md (hunk 6 above)
Comment (prayank23) [reply doc/i2p.md:16]: https://github.com/bitcoin/bitcoin/pull/22250#discussion_r654495784

File: doc/i2p.md (hunk 5 above)
Comment (prayank23) [reply doc/i2p.md:72]: It looks correct but we can add/remove information based on how issue in other PR is resolved and merged.

**Comment by prayank23** [comment]
Should we add information about ports in this doc? https://github.com/bitcoin/bitcoin/issues/22634#issuecomment-893834628

File: doc/i2p.md (hunk 6 above)
Comment (jonatack) [reply doc/i2p.md:16]: Thanks for the info! ISTM the link should still be provided.

**Comment by jonatack** [comment]
Your suggestion? Feel free to propose a text.

**Comment by prayank23** [comment]
Looks like we need all ports open for outbound: https://www.reddit.com/r/i2p/comments/fzwwd0/confused_on_what_ports_to_open/fn76ly9

File: doc/i2p.md (hunk 6 above)
Comment (Rspigler) [reply doc/i2p.md:16]: I tested v22.0rc2 with the official Java i2prouter with no issues on Debian.  You have to follow the instructions at their download link, and then visit 127.0.0.1:7657 to start SAM.  You also need to type `i2prouter start` in terminal.

Comment (Rspigler) [reply doc/i2p.md:16]: 127.0.0.1:7657 -> ConfigureHomepage -> Clients  then hit the Play Button next to "SAM application Bridge".  On the left side of the page, there should be a green light next to "Shared Clients".

File: doc/i2p.md [hunk 7]
```diff
@@ -10,11 +10,22 @@ started with I2P terminology.
...
+- [i2pd (I2P Daemon)](https://github.com/PurpleI2P/i2pd)
+  ([documentation](https://i2pd.readthedocs.io/en/latest)), a lighter
+  alternative in C++ (successfully tested with version 2.23 and up; version 2.36
+  or later recommended)
+- [i2p-zero](https://github.com/i2p-zero/i2p-zero)
+- [other alternatives](https://en.wikipedia.org/wiki/I2P#Routers)
+
```
Comment (Rspigler) [inline doc/i2p.md:23]: For the official i2prouter implementation in Java, visit their [download page](https://geti2p.net/en/download) and follow the instructions for your Operating System.

Once installed, open a terminal window and type `i2prouter start`.  Then visit `127.0.0.1:7657` in your browser to enable SAM.  To do so, select: "Configure Homepage", then "Clients", and finally select the "Play Button" next to `SAM application Bridge`. On the left side of the page, there should be a green light next to "Shared Clients".

**Comment by Rspigler** [comment]
@prayank23 I don't believe that is correct.  I am behind a router/firewall I don't control right now (haven't forwarded any ports) and I am able to successfully create an i2p address (`ugmw4zuyhmrm7anfctjamcjgi32daor4ulqdrxtn66owgn2oyebq.b32.i2p:0`) and connect out to other i2p nodes (although they are not very long lasting connections, and I have no inbound connections despite setting `i2pacceptincoming=1`).

That being said, I also have no incoming tor connections to my HS (`kwn76yxlsaucsbywm5sy4q5ebo6atao5escejchwjkzu2l5lwkkkzdqd.onion:8333`) which I know doesn't require any port forwarding, so I'm not sure what's going on.  I have been running my node for many hours.

`bitcoin.conf`:

> 
> listen=1
> assumevalid=0
> dbcache=1000
> addresstype=bech32
> avoidpartialspends=1
> changetype=bech32
> maxapsfee=0.01
> i2psam=127.0.0.1:7656
> debug=tor
> debug=i2p
> i2pacceptincoming=1
> 

In my debug file, I see that my node is trying to connect to lots of i2p nodes, but I have tons of connection timeout errors.

**Comment by prayank23** [comment]
> Edit: Maybe you could try to connect to me?

Sure. Will experiment tomorrow. Will ping you in IRC. It's 4 AM here right now 😴

File: doc/i2p.md (hunk 6 above)
Comment (jonatack) [reply doc/i2p.md:16]: Thanks for the info, @Rspigler. Resolving here for now.

**Comment by jonatack** [comment]
@Rspigler I manually connected successfully to both of your addresses.

File: doc/i2p.md (hunk 5 above)
Comment (jonatack) [reply doc/i2p.md:72]: Ok. Resolving for now.

**Comment by Rspigler** [comment]
I see your inbound for both i2p and tor, so I think we should be good, unless other users start reporting high ping times for these networks with v22.0.

Thanks for helping!

**Comment by prayank23** [comment]
> I see your inbound for both i2p and tor, so I think we should be good, unless other users start reporting high ping times for these networks with v22.0.

Things that you should try:

1. Check firewall status: `sudo ufw status verbose`
2. Deny incoming traffic, restart and see if i2p works fine in Core for incoming peers
3. Block some ports for outgoing and see if i2p works for outbound peers in Core
4. Results for `ss nlt` and `nmap`

File: doc/i2p.md (hunk 7 above)
Comment (RandyMcMillan) [reply doc/i2p.md:23]: If the user has successfully config'd to this point - a link will take the user to the correct page.

Maybe add a link for convenience?

Then visit [127.0.0.1:7657/configclients](http://127.0.0.1:7657/configclients)

**Comment by josedanielrojaspixel** [comment]
Enviado desde mi HUAWEI Mate 20 Pro-------- Mensaje original --------De: ***@***.***" ***@***.***>Fecha: lun., 9 ago. 2021 4:35 p. m.Para: bitcoin/bitcoin ***@***.***>CC: Subscribed ***@***.***>Asunto: Re: [bitcoin/bitcoin] doc, test: improve i2p/tor docs and i2p reachable unit tests (#22648)
@RandyMcMillan commented on this pull request.



In doc/i2p.md:
> -enabled is required (there is an [official one](https://geti2p.net) and
-[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP
-address and port the SAM proxy is listening to; usually, it is
-`127.0.0.1:7656`. Once it is up and running with SAM enabled, use the following
-Bitcoin Core options:
+
+enabled is required. Options include:
+
+- [i2p-router](https://geti2p.net), the official implementation in Java
+- [i2pd (I2P Daemon)](https://github.com/PurpleI2P/i2pd)
+  ([documentation](https://i2pd.readthedocs.io/en/latest)), a lighter
+  alternative in C++ (successfully tested with version 2.23 and up; version 2.36
+  or later recommended)
+- [i2p-zero](https://github.com/i2p-zero/i2p-zero)
+- [other alternatives](https://en.wikipedia.org/wiki/I2P#Routers)
+

If the user has successfully config'd to this point - a link will take the user to the correct page.
Maybe add a link for convenience?
Then visit 127.0.0.1:7657/configclients

File: doc/i2p.md (hunk 7 above)
Comment (josedanielrojaspixel) [reply doc/i2p.md:23]: L09

Comment (jonatack) [reply doc/i2p.md:23]: It's not clear to me if a change is being requested here, and what, precisely.  Install how-to seems a bit redundant with the existing docs of the various implementations.

**Comment by jonatack** [comment]
Hm, more than 30 comments and people asking questions, no approvals of this change.

**Comment by jonatack** [comment]
Hm, more than 30 comments and people asking questions, but no approvals of this change.

**Review by ghost — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/22648/commits/e3a58140dc8fe20b76d5ef38074e37bca7111b79

nit: 

1. Would be great if we can add things suggested by @Rspigler in https://github.com/bitcoin/bitcoin/pull/22648/#discussion_r684835687

2. And information about ports or at least this link: https://geti2p.net/en/faq#ports

I have added both things in https://github.com/BlockchainCommons/Learning-Bitcoin-from-the-Command-Line/pull/408

For ports I tried experimenting with allow/deny in `ufw` and see if it affects outgoing and incoming connections in Bitcoin Core. 

TL;DR: 

```
Ports required by i2p:

1. Outbound (Internet facing): a random port between 9000 and 31000 is selected however its better if all ports are open and it doesn't affect your security. You can check firewall status using `sudo ufw status verbose` which shouldn't deny outgoing by default.

2. Inbound (Internet facing): Optional

Local ports: https://geti2p.net/en/faq#ports
```

If something looks incorrect feel free to comment :)

File: doc/i2p.md (hunk 7 above)
Comment (Rspigler) [reply doc/i2p.md:23]: I think adding the [above](https://github.com/bitcoin/bitcoin/pull/22648#discussion_r684835687) at Line 23 (insert new line first) would be helpful.  I agree instructions for every implementation would be overkill, but I was completely new to i2p, and I had to do a bit of digging to figure out how to set it up.  I found the wording a bit confusing (I thought I had to install a router and install SAM).  The documentation on the website is mostly about setting up a browser.  I think adding instructions for users for the reference implementation could be helpful in increasing i2p nodes.  The tor docs include a bit of instructions as well

**Comment by Rspigler** [comment]
> a random port between 9000 and 31000

I can confirm this

File: doc/i2p.md (hunk 7 above)
Comment (jonatack) [reply doc/i2p.md:23]: Thanks. I agree it might be helpful, but I don't know if those instructions are platform agnostic. We would also need to maintain it / keep it up to date. Most people I talk to are using i2pd (admittedly perhaps because it is very easy to install), so if we explain how to install i2p-router, we should probably also do it for i2pd.  Sounds like a dedicated "How to Install an I2P Router" section.

In any case, feel free to propose this separately. I would rather not make this pull larger.

**Comment by jonatack** [comment]
Thanks for the suggestions. I'd rather not copy information here that is already in https://geti2p.net/en/faq#ports, as it would add maintainance here and could go out of date.

**Comment by jonatack** [comment]
Thanks @prayank23! And thanks for the suggestions. I'd rather not copy information here that is already in https://geti2p.net/en/faq#ports, as it would add maintainance here and could go out of date.

File: src/test/net_tests.cpp [hunk 8]
```diff
@@ -771,18 +771,22 @@ BOOST_AUTO_TEST_CASE(LimitedAndReachable_Network)
...
 
     SetReachable(NET_IPV4, true);
     SetReachable(NET_IPV6, true);
     SetReachable(NET_ONION, true);
+    SetReachable(NET_I2P, true);
 
     BOOST_CHECK_EQUAL(IsReachable(NET_IPV4), true);
     BOOST_CHECK_EQUAL(IsReachable(NET_IPV6), true);
     BOOST_CHECK_EQUAL(IsReachable(NET_ONION), true);
+    BOOST_CHECK_EQUAL(IsReachable(NET_I2P), true);
```
Comment (vasild) [inline src/test/net_tests.cpp:789]: nit:
```cpp
BOOST_CHECK_EQUAL(foo, true);
BOOST_CHECK_EQUAL(bar, false);
// vs
BOOST_CHECK(foo);
BOOST_CHECK(!bar);
```

but the surrounding code is already using the former, so it is ok to leave it as it is.

**Review by vasild — APPROVED** [review]
ACK e3a58140dc8fe20b76d5ef38074e37bca7111b79

**Comment by Rspigler** [comment]
ACK e3a58140dc8fe20b76d5ef38074e37bca7111b79

File: doc/i2p.md [hunk 9]
```diff
@@ -10,11 +10,22 @@ started with I2P terminology.
 ## Run Bitcoin Core with an I2P router (proxy)
 
 A running I2P router (proxy) with [SAM](https://geti2p.net/en/docs/api/samv3)
-enabled is required (there is an [official one](https://geti2p.net) and
-[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP
-address and port the SAM proxy is listening to; usually, it is
-`127.0.0.1:7656`. Once it is up and running with SAM enabled, use the following
-Bitcoin Core options:
+
```
Comment (laanwj) [inline doc/i2p.md:13]: extra newline?

Comment (jonatack) [reply doc/i2p.md:13]: oops

File: src/test/net_tests.cpp (hunk 8 above)
Comment (jonatack) [reply src/test/net_tests.cpp:789]: Updated the surrounding ones (and saw a missing NET_I2P assertion, added). Thanks!

File: doc/i2p.md (hunk 9 above)
Comment (jonatack) [reply doc/i2p.md:13]: removed extra newline, good catch!

**Comment by jonatack** [comment]
Updated `git diff e3a5814 0175977` per review feedback (thanks!) and corrected the name of the first I2P router from `i2p-router` to `i2prouter (I2P Router)`, which is also more similar to the second router `i2pd (I2P Daemon)`.

**Comment by Rspigler** [comment]
Re-ACK 017597767b977bb8fe11be8e2012130df1dffd30

**Review by vasild — APPROVED** [review]
ACK 017597767b977bb8fe11be8e2012130df1dffd30

**Comment by prayank23** [comment]
reACK https://github.com/bitcoin/bitcoin/commit/017597767b977bb8fe11be8e2012130df1dffd30

Changes since last review: https://github.com/bitcoin/bitcoin/pull/22648#issuecomment-901964610

File: doc/tor.md [hunk 10]
```diff
@@ -57,11 +57,11 @@ outgoing connections, but more is possible.
...
-                    ipv4, ipv6 or onion. If you use this option with values other
-                    than onion you *cannot* disable onion connections; outgoing onion
-                    connections will be enabled when you use -proxy or -onion. Use
-                    -noonion or -onion=0 if you want to be sure there are no outbound
-                    onion connections over the default proxy or your defined -proxy.
+                    onlynet=ipv4, onlynet=ipv6, onlynet=onion, onlynet=i2p.
+                    Warning: if you use -onlynet with values other than onion, and
+                    the -onion or -proxy option is set, then outgoing onion
+                    connections will still be made; use -noonion or -onion=0 to
+                    disable outbound onion connections in this case.
```
[the remark this answers is not in the document]
Comment (vasild) [reply doc/tor.md:64]: > Can't we make `-onlynet=ipv4` ... do as one would expect ...

Done in https://github.com/bitcoin/bitcoin/pull/22834.
